zipkin-storage/elasticsearch/src/main/java/zipkin2/elasticsearch/internal/JsonReaders.java

```diff
@@ -25,85 +25,77 @@
 public final class JsonReaders {
 
   /**
-   * This saves you from having to define nested types to read a single value
-   *
-   * <p>Instead of defining two types like this, and double-checking null..
-   *
-   * <pre>{@code
-   * class Response {
-   *   Message message;
-   * }
-   * class Message {
-   *   String status;
-   * }
-   * JsonAdapter<Response> adapter = moshi.adapter(Response.class);
-   * Message message = adapter.fromJson(body.source());
-   * if (message != null && message.status != null) throw new IllegalStateException(message.status);
-   * }</pre>
-   *
-   * <p>You can advance to the field directly.
+   * Navigates to a field of a JSON-serialized object. For example,
    *
    * <pre>{@code
-   * JsonReader status = enterPath(JsonReader.of(body.source()), "message", "status");
+   * JsonReader status = enterPath(JsonAdapters.jsonParser(stream), "message", "status");
    * if (status != null) throw new IllegalStateException(status.nextString());
    * }</pre>
    */
   @Nullable
-  public static JsonReader enterPath(JsonReader reader, String path1, String path2)
+  public static JsonParser enterPath(JsonParser parser, String path1, String path2)
       throws IOException {
-    return enterPath(reader, path1) != null ? enterPath(reader, path2) : null;
+    return enterPath(parser, path1) != null ? enterPath(parser, path2) : null;
   }
 
   @Nullable
-  public static JsonReader enterPath(JsonReader reader, String path) throws IOException {
-    try {
-      if (reader.peek() != JsonReader.Token.BEGIN_OBJECT) return null;
-    } catch (EOFException e) {
-      return null;
-    }
-    reader.beginObject();
-    while (reader.hasNext()) {
-      if (reader.nextName().equals(path) && reader.peek() != JsonReader.Token.NULL) {
-        return reader;
+  public static JsonParser enterPath(JsonParser parser, String path) throws IOException {
+    if (!parser.isExpectedStartObjectToken()) return null;
+    JsonToken value;
+    while ((value = parser.nextValue()) != JsonToken.END_OBJECT) {
+      if (value == null) {
+        // End of input.
+        throw new IOException("End of input while parsing object.");
```

Comment:
our old code didn't throw I think, rather returned null

Reply:
I believe this essentially corresponds to `reader.endObject()` - Moshi will generally throw JsonDataException runtime exceptions on malformed JSON or EOF within one of its methods. I stuck with throwing IOException since all our methods declare throwing IOException but can change to IllegalArgumentException if it makes more sense. Returning `null` will probably not reproduce the old behavior.